### PR TITLE
feat(overlay): detach on outside click

### PR DESCRIPTION
**Pull request by wnvko**
Allow detaching of the overlay via `detachOnOutsideClick` property in `OverlayConfig`. If set to true the overlay will determine if user performed outside click. If so overlay will detach itself.

Additionally `excludeFromOutsideClick` property is added to `OverlayConfig`. This is an array of `HTMLElements`. Developer may add elements to this array. If user clicks on any of these elements the click will not be considered outside click.

### CODE REVIEW

File: src/cdk/overlay/overlay-config.ts [hunk 1]
```diff
@@ -59,6 +59,16 @@ export class OverlayConfig {
...
 
+  /**
+   * Whether the overlay should be detached on outside click
+   */
+  detachOnOutsideClick?: boolean = false;
+
+  /**
+   * Array of HTML elements clicking on which should not be considered as outside click
+   */
+  excludeFromOutsideClick?: HTMLElement[] = [];
```
Comment (crisbeto) [inline src/cdk/overlay/overlay-config.ts]: I'm not quite sure what the use case is for this.

File: src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import {Inject, Injectable, OnDestroy} from '@angular/core';
+import {OverlayRef} from '../overlay-ref';
+
+/**
+ * Service for dispatching mouse click events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+@Injectable({providedIn: 'root'})
+export class OverlayMouseClickDispatcher implements OnDestroy {
```
Comment (crisbeto) [inline src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts]: It looks like this class shares most of its code with the keyboard dispatcher. I think that we should only have one dispatcher that deals with both so we don't have to duplicate all of the code. If we rolled this into the keyboard dispatcher the class name won't make sense anymore, but we can rename it in a backwards-compatible way.

File: src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      });
+
+      // If it is inside click just return - we should do nothing
+      // If it is outside click check if we should detach overlay and proceed
+      // with the next overlay
+      if (isInsideClick) {
+        return;
+      } else {
+        if (config.detachOnOutsideClick) {
+          overlayRef.detach();
```
Comment (crisbeto) [inline src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts]: Having this call happen here feel a little weird, because all of the other similar functionality happens inside the `OverlayRef`. I think that we should follow the same pattern where the dispatcher just emits some kind of event and then it's up to the overlay to decide what it wants to do based on its options.

**Comment by wnvko** [comment]
> On top of the code comments, I spent some time testing it against an iOS device and it looks like this approach has the same issue that I ran into when I tried doing this a couple of years ago and which was raised in the design doc: touch devices won't dispatch mouse events when the user is tapping on a non-clickable element. This is a problem because the user can't tap anywhere outside of the overlay, like they can now, to close the element. The recording is from the iOS Simulator from xcode, but I'm seeing the same issue on a device running iOS 12.3.1. Note how tapping anywhere on the body doesn't close it, but it works once I tap on a clickable element like one of the buttons.
> 
> Aug-07-2019 11-28-02

I can solve this by forcing the `documentElement` to be "clickable" under iOS. This could be achieved by adding `onclick` attribute or by setting the `cursor` style to `pointer`. I can check if the OS is iOS and set the cursor. Of course, when all the overlays get closed I will restore the original cursor style. What do you think about such a fix?

**Comment by crisbeto** [comment]
I don't know whether this is only an issue on iOS or whether we'd need to do this for all touch devices. If we end up having to do it for all, it could look weird on hybrid devices that both have a touch screen and a mouse. What do you think @jelbourn?

**Comment by jelbourn** [comment]
Yes, we would definitely have to confirm that this works for all touch devices. I have a pixelbook I can test if you make the change. I would go with adding a no-op event listener to the document outside the zone if that makes the browser see all of the content as "clickable". The listener should only be applied while the overlay is open.

File: src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts (hunk 2 above)
Comment (jelbourn) [reply src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts]: Agreed that, if they share a lot of code, we should combine the two classes. This name also says "Mouse"- does this apply for touch event as well?

File: src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts (hunk 3 above)
Comment (jelbourn) [reply src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts]: Agreed, it should be outside the responsibility of the event manager to actually dismiss the overlay; the overlay should instead respond to a signal from the keyboard manager.

File: src/cdk/overlay/overlay-config.ts (hunk 1 above)
Comment (wnvko) [reply src/cdk/overlay/overlay-config.ts]: This array will contain elements which will be excluded from outside clicks. Let say you have two buttons. Clicking on each one will open a mat-menu. Right now if you click on button one mat-menu will be shown. Then if you click on button two mat-menu will be closed, but second mat-menu will not be shown. As described in [this](https://github.com/angular/components/issues/9320) issue.
If we set `detachOnOutsideClick` for mat-menu to `true` this will work (as well as set `hasBackdrop` to false). However in this case when you click on button one mat-menu will be shown. If at this point you click again at button one this will happen:
1. Document's body will recognize this as outside click and will force mat-menu to close.
2. Click event will reach the button one. At this point the mat-menu is already closed, so the button will try to open it.
To prevent this from happen we should say to overlay consider clicking on the button as not outside click. And we can achieve this by adding the  button to `excludeFromOutsideClick` array.

File: src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts (hunk 2 above)
Comment (wnvko) [reply src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts]: @crisbeto and @jelbourn you are right. I was looking at `OverlayKeyboardDispatcher` when I was implementing `OverlayClickDispatcher`. However looking at the code if we combine both these in one single class it will be a little awkward. On `add` `OverlayKeyboardDispatcher` adds event listener for `keydown` while `OverlayClickDispatcher` adds one for `click`. Then in the event listener function both dispatchers implements different logic. If I put both in one dispatcher it should have a lot of `ifs` inside. Additionally if in future a need for another dispatcher arise we should add more `ifs`. This is why I’ve extracted the common logic in base `OverlayBaseDispatcher` class which both keyboard and click dispatchers extend. What you think about this?

File: src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts (hunk 3 above)
Comment (wnvko) [reply src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts]: I've fixed this. Now new `OverlayClickDispatcher` fires `_clickEvents`. Anyone interested should subscribe to the public `clickEvents` observable in the overlay ref.

**Comment by wnvko** [comment]
@crisbeto and @jelbourn regarding the iOS issue I am setting now the body's cursor style to `pointer` when `overlayRef` is added. As soon as last `overlayRef` is removed I am restoring original style.
I've tested this on iPad and iPhone both with iOS 12.3.1 and it works.
I also have tested this on Android 9 device, on my PC and on PC with touch screen - it was ok on all these devices.

**Comment by jelbourn** [comment]
I tried this out locally just now- I set `detachOnOutsideClick: true` in the `connected-overlay-demo`, but clicking outside the panel doesn't dismiss the overlay. Am I missing something?

**Comment by wnvko** [comment]
> I tried this out locally just now- I set `detachOnOutsideClick: true` in the `connected-overlay-demo`, but clicking outside the panel doesn't dismiss the overlay. Am I missing something?

Now if `detachOnOutsideClick` is set to `true` `OverlayRef`'s `clickEvents` will dispatch event on outside click. You should add, in `connected-overlay-demo` in `openWithConfig` something like this:
```
    this.overlayRef.clickEvents().subscribe(() => {
      this.close();
    });
```
In this way this is now working similarly to `OverlayKeyboardDispatcher`

**Comment by jelbourn** [comment]
I see; I think the naming is currently misleading, then. `detachOnOutsideClick` would make me think the overlay will automatically do this. It's also not totally clear that `clickEvents` would include clicks outside the overlay, since the description says "observable of keydown events targeted to this overlay".

Would it make more sense to just have a stream of `outsideClicks` (similar to backdropClick) that the user can handle in the same manner?

**Comment by wnvko** [comment]
The initial idea was setting the `detachOnOutsideClick` to automatically do that. Then we switched it to observable, which is better I think. As a result `detachOnOutsideClick` became obsolete. I will remove it now. I should update and the design doc, should not I?

**Comment by jelbourn** [comment]
That sounds good

File: src/cdk/overlay/dispatchers/overlay-click-dispatcher.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    // the loop.
+    for (let i = overlays.length - 1; i > -1; i--) {
+      const overlayRef: OverlayRef = overlays[i];
+      if (overlayRef._clickEventSubscriptions < 1) {
+        continue;
+      }
+
+      const config = overlayRef.getConfig();
+      const elements = [...config.excludeFromOutsideClick!];
+      elements.push(overlayRef.overlayElement);
```
Comment (jelbourn) [inline src/cdk/overlay/dispatchers/overlay-click-dispatcher.ts]: I would name this `excludedElements`. You can also omit the push and include the element in the array directly:
```ts
const excludedElements = [...config.excludeFromOutsideClick, overlayRef.overlayElement];
```

File: src/cdk/overlay/overlay-config.ts (hunk 1 above)
Comment (jelbourn) [reply src/cdk/overlay/overlay-config.ts]: Now that `detachOnOutsideClick` has been removed in favor of having a click stream, is this option still necessary? Consumers of the API can now make the determination of what to do on the click based on what was clicked, right?

File: src/cdk/overlay/overlay-directives.ts [hunk 5]
```diff
@@ -222,6 +222,9 @@ export class CdkConnectedOverlay implements OnDestroy, OnChanges {
   /** Emits when there are keyboard events that are targeted at the overlay. */
   @Output() overlayKeydown = new EventEmitter<KeyboardEvent>();
 
+  /** Emits when there are mouse events that are targeted at the overlay. */
+  @Output() overlayClick = new EventEmitter<MouseEvent>();
```
Comment (jelbourn) [inline src/cdk/overlay/overlay-directives.ts]: Is this clicks _inside_ the overlay or _outside_ the overlay? If it's the latter, I would rename this to `outsideClick`

File: src/cdk/overlay/dispatchers/overlay-base-dispatcher.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import {Inject, OnDestroy} from '@angular/core';
+import {OverlayRef} from '../overlay-ref';
+
+
+/**
+ * Service for dispatching keyboard events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+export abstract class OverlayBaseDispatcher implements OnDestroy {
```
Comment (jelbourn) [inline src/cdk/overlay/dispatchers/overlay-base-dispatcher.ts]:
```suggestion
export abstract class BaseOverlayDispatcher implements OnDestroy {
```

File: src/cdk/overlay/overlay-ref.ts [hunk 7]
```diff
@@ -268,6 +305,11 @@ export class OverlayRef implements PortalOutlet, OverlayReference {
     return this._keydownEventsObservable;
   }
 
+  /** Gets an observable of mouse outside click events targeted to this overlay. */
+  clickEvents(): Observable<MouseEvent> {
```
Comment (jelbourn) [inline src/cdk/overlay/overlay-ref.ts]: Is this clicks _inside_ or _outside_ the overlay?

File: src/cdk/overlay/dispatchers/overlay-click-dispatcher.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  /** Click event listener that will be attached to the body propagate phase. */
+  private _clickListener = (event: MouseEvent) => {
+    const overlays = this._attachedOverlays;
+
+    // Dispatch the mouse event to the top overlay which has subscribers to its mouse events.
+    // We want to target all overlays for which the click could be considered as outside click.
+    // As soon as we reach an overlay for which the click is not outside click we break off
+    // the loop.
+    for (let i = overlays.length - 1; i > -1; i--) {
+      const overlayRef: OverlayRef = overlays[i];
```
[the remark this answers is not in the document]
Comment (wnvko) [reply src/cdk/overlay/dispatchers/overlay-click-dispatcher.ts]: Thank you for the suggestion - added in my commit with other changes

File: src/cdk/overlay/overlay-ref.ts (hunk 7 above)
Comment (wnvko) [reply src/cdk/overlay/overlay-ref.ts]: definitely `outsideClickEvents`

File: src/cdk/overlay/overlay-directives.ts (hunk 5 above)
Comment (wnvko) [reply src/cdk/overlay/overlay-directives.ts]: Regarding `excludeFromOutsideClick` I was thinking for the situations where `overlayRef` has no clue who opened it, and one could not determine in the handler of `outsideClickEvents`, clicking on which element on the page should or should not detach the overlay.

File: src/cdk/overlay/overlay-config.ts (hunk 1 above)
Comment (wnvko) [reply src/cdk/overlay/overlay-config.ts]: I was thinking for the situations where `overlayRef` has no clue who opened it, and one could not determine in the handler of `outsideClickEvents`, clicking on which element on the page should or should not detach the overlay.

File: src/cdk/overlay/public-api.ts [hunk 9]
```diff
@@ -10,14 +10,14 @@ export * from './overlay-config';
 export * from './position/connected-position';
 export * from './scroll/index';
 export * from './overlay-module';
+export * from './dispatchers/index';
```
Comment (jelbourn) [inline src/cdk/overlay/public-api.ts]: Could we actually exclude `BaseOverlayDispatcher` from the public API?

File: src/cdk/overlay/overlay-ref.ts [hunk 10]
```diff
@@ -268,6 +305,11 @@ export class OverlayRef implements PortalOutlet, OverlayReference {
     return this._keydownEventsObservable;
   }
 
+  /** Gets an observable of mouse outside click events targeted to this overlay. */
+  outsideClickEvents(): Observable<MouseEvent> {
```
[the remark this answers is not in the document]
Comment (wnvko) [reply src/cdk/overlay/overlay-ref.ts]: With current implementation I cannot change the observable from `Observable<MouseEvent>` to  `Observable<PointerEvent>`  as I am handling `click` event in the dispatcher. Should I try to switch to pointer event in the dispatcher, even know this is not supported on iOS pre 13.2?

[the remark this answers is not in the document]
Comment (jelbourn) [reply src/cdk/overlay/overlay-ref.ts]: MouseEvent is probably fine, then

**Comment by jelbourn** [comment]
Looks like `BaseOverlayDispatcher` still needs to be excluded from the public api

**Comment by wnvko** [comment]
> Looks like `BaseOverlayDispatcher` still needs to be excluded from the public api

I've already removed it in this [commit](https://github.com/angular/components/pull/16611/commits/a375a9e26f43b7160567baff1a3d82ec1812d65d) :)

**Comment by jelbourn** [comment]
Looks good, just needs a rebase

**Comment by wnvko** [comment]
@jelbourn I had some circular dependencies issues. To solve them I have updated `OverlayReference`, as well as have removed one circular dependency from `goldens/ts-circular-deps.json`. I was wondering is this the right way, or should I leave `OverlayReference` as it was and add some more dependencies in `goldens/ts-circular-deps.json`.

Also one of the checks is failing. I think this is not because of my changes.

**Comment by jelbourn** [comment]
Thanks for rebasing! Removing circular deps is always good. I ran this PR through Google's internal presubmit system last night and it looks like there are at least ~20 failures. I'll try to investigate those failures in the next week.

**Comment by andrewseguin** [comment]
@wnvko Thanks for having patience with us on this PR - we really appreciate the contribution and this change will actually be a big help for one of our upcoming features with menu. Would you mind rebasing it once again, and I'll prioritize getting this change in?
